portable/GCC/ARM_CRx_MPU/mpu_wrappers_v2_asm.S

```diff
@@ -0,0 +1,1334 @@
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/*---------------------------------------------------------------------------*/
+
+    .arm
+    .syntax unified
+    .section freertos_system_calls
+#define FREERTOS_ASSEMBLY
+    #include "FreeRTOSConfig.h"
+    #include "portmacro_asm.h"
+    #include "mpu_syscall_numbers.h"
+#undef FREERTOS_ASSEMBLY
+
+/* ------------------ Start of Port Specific System Calls ------------------ */
+
+/**
+ * Function: void vPortYield
+ * Inputs:      VOID
+*/
+.align 4
+.global vPortYield
+.type vPortYield, %function
+vPortYield:
+    /* Make the SVC to swap tasks */
+    SVC     #portSVC_YIELD
+    /* After yielding to another task, resume executing the calling task */
+    BX  LR
+
+
+/*-------------------------------------------------------------------------------*/
+/* vPortSystemCallExit */
+.align 4
+.global vPortSystemCallExit
+.type vPortSystemCallExit, %function
+vPortSystemCallExit:
+    /* Make the SVC to exit a system call */
+    SVC     #portSVC_SYSTEM_CALL_EXIT
+    /* After performing the requested FreeRTOS API, return to the calling task */
+    BX      LR
+
+/*-------------------------------------------------------------------------------*/
+/**
+ * Function: BaseType_t xPortIsPrivileged
+ * Inputs:      VOID
+*/
+.align 4
+.weak xPortIsPrivileged
+.type xPortIsPrivileged, %function
+xPortIsPrivileged:
+    /* Load value of SPSR into R0 */
+    MRS     R0, CPSR
+    /* Get the relevant mode bits */
+    AND     R0, R0, #0x1F
+    /* Check if we're in user mode */
+    CMP     R0, #USER_MODE
+    /* If we are in user mode set R0 to 0 for the return */
+    MOVEQ   R0, #0x0
+    /* Otherwise set R0 to 1 for the return */
+    MOVNE   R0, #0x01
+    BLX      LR
+
+/*-------------------------------------------------------------------------------*/
+/**
+ * Function: UBaseType_t ulPortCountLeadingZeros
+ * Inputs:      UBaseType_t ulBitmap
+*/
+.align 4
+.weak ulPortCountLeadingZeros
+.type ulPortCountLeadingZeros, %function
+ulPortCountLeadingZeros:
+    /* Count the leading zeros and return in R0 */
+    CLZ     R0, R0
+    BX      LR
+
+/* ------------------- End of Port Specific System Calls ------------------- */
+
+.macro processorInUserMode
+    /* Push R0 before using it */
+    PUSH    {R0}
+    /* Load value of SPSR into R0 */
+    MRS     R0, CPSR
+    /* Get the relevant mode bits */
+    AND     R0, R0, #0x1F
+    /* Determine if the CPU is in user mode */
+    CMP     R0, #USER_MODE
+    /* Restore the pushed register */
+    POP     {R0}
+.endm
+
+/*---------------------------------------------------------------------------*/
+
+.extern MPU_xTaskGetTickCountImpl
+/**
+ * Function: TickType_t MPU_xTaskGetTickCount
+ * Inputs:     void - No Inputs
+*/
+.align 4
+.global MPU_xTaskGetTickCount
+.type MPU_xTaskGetTickCount, function
+MPU_xTaskGetTickCount:
+    processorInUserMode
+    SVCEQ   #SYSTEM_CALL_xTaskGetTickCount
+    B       MPU_xTaskGetTickCountImpl
```

Comment:
I am probably misreading this - but if the process in usermode then `processInUserMode` will set the condition status field to true which will in turn result in the `SVCEQ` call executing followed by the unconditional branch which calls the impl.

What happens here if you're in system mode? From what I can tell it wouldn't log the system call but would still take the branch. Is that the correct behavior?

Reply:
Yes that is correct. When the processor is in [User Mode](https://developer.arm.com/documentation/ddi0406/cb/System-Level-Architecture/The-System-Level-Programmers--Model/ARM-processor-modes-and-ARM-core-registers/ARM-processor-modes?lang=en) it will make a SVC, which will route through the [portASM.S:FreeRTOS_SVC_Handler](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/portASM.S#L288) to [portASM.S:svcSystemCallEnter](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/portASM.S#L397)

When an unprivileged task goes through [portASM.S:svcSystemCallEnter](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/portASM.S#L397) the stack pointer of the task is temporarily moved to [ulSystemCallStackBuffer](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/portmacro.h#L628), and the task's link register is set to [vPortSystemCallExit](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/mpu_wrappers_v2_asm.S#L61):

```
    /* Store the task's SP and LR to xSYSTEM_CALL_STACK_INFO */
    STM     R11, {R13-R14}^
    /* It's a compilation warning to use the ! for writeback, so manually move R11 */
    ADD     R11, R11, 0x8
    /* Load pulSystemCallStackPointer and pulSystemCallLinkRegister now */
    LDM     R11, {R13-R14}^
```

The handler also loads the function from the table declared in [portable/Common/mpu_wrappers_v2](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V11.0.1/portable/Common/mpu_wrappers_v2.c#L4896):

```
    /* Load the base address of the uxSystemCallImplementations[] table into R14 */
    LDR     R14, =uxSystemCallImplementations

    /** Shift the value of R11, the SVC number, left by two to get the jump offset
     * Add this offset to R14, which holds the jump table address. This is the address
     * of the SVC that the relevant function is trying to complete.
     * Now when the Link Register is loaded as the Program Counter at the end of this
     * handler, the caller will immediately execute the requested function */
    LDR     R14, [R14, R11, lsl #2]
```

Lastly the task is set to operate in SYSTEM_MODE for the duration of the [FreeRTOS System Call](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/portASM.S#L432C1-L439C25):

```
    /* Swap the SPSR to SYS_MODE for the System Call. Move SPSR into R12 */
    MRS     R12, SPSR
    /* Set the MODE bits to SYS_MODE */
    ORR     R12, R12, #SYS_MODE
    /* Assign the new value to SPSR */
    MSR     SPSR_cxsf, R12
    /* Leave through the SVC Exit */
    B   SVC_Handler_Exit
```

This way if an unprivileged task makes a FreeRTOS System Call all information related to that call is kept in privileged data. Then at the end of the system call the function will pop the LR off and exit through [vPortSystemCallExit](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/d965464f4e2a22d5bcd0d85ae63fac4356ff8b8d/portable/GCC/ARM_CRx_MPU/mpu_wrappers_v2_asm.S#L61). 

If a caller is already in a privileged operating mode they will fail this check and branch directly to the function, as they are already privileged and have the ability to execute the FreeRTOS-Kernel functions placed in [PRIVILEGED_FUNCTIONS](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V11.0.1/include/mpu_wrappers.h#L270)

Reply:
Thank you for the detailed explanation here.